search/query/date_range.go

```diff
@@ -41,6 +41,14 @@ type BleveQueryTime struct {
 	time.Time
 }
 
+var minValidTime time.Time
+var maxValidTime time.Time
+
+func init() {
+	minValidTime, _ = time.Parse(time.RFC3339, "1677-12-01T00:00:00Z")
+	maxValidTime, _ = time.Parse(time.RFC3339, "2262-04-11T11:59:59Z")
+}
```

Comment:
What do you think of this approach to finding the limits, it avoids parsing/errors and (I think) gets the exact end points we could support?

https://play.golang.org/p/Uz5m5i7CZZR

Reply:
Yea this looks good, but looks like the values that the playground code generates are a bit off. This is the output I see ...

```
epoch: 1970-01-01T00:00:00Z
max: 2262-04-11T23:47:16Z
min: 1677-09-21T00:12:43Z
```

When I plug in these values for minValidTime, maxValidTime, I see an overflow. 
https://play.golang.org/p/e7tAiuN8i7I

Am I doing something wrong here?

Reply:
No, it turns out that internally time is tracked in two variables, so limits are not quite as straightforward.  See this question for some of the details:

https://stackoverflow.com/questions/25065055/what-is-the-maximum-time-time-in-go

I guess I'll back off having some elegant way to compute these.

But I now have a new requirement, we should export those min/max values.  The reason is that we have examples of users try to express the query, "has a date/time indexed".  Previously this could be done with [zeroTime+1min, zeroTime).  Given that now there is some enforced restriction on the values, we should give users a symbolic way to express the same thing:

[minTime, zeroTime)

Reply:
Actually I now understand why the min overflowed in your example.  My way of initializing things went directly to the nanoseconds.  However, your version starts with the text represenation, without the nanosecond precision.  This rounds down to 0 nanoseconds, which is further in the negative direction from the epoch.

Here is a program (have to run locally since go playground is 32-bit) which attempted to binary search finding the first overflow situation:

```
package main

import (
	"fmt"
	"math"
	"sort"
	"time"
)

func main() {
	epoch := time.Unix(0, 0)
	epochUnixNano := epoch.UnixNano()
	fmt.Printf("epoch unix nano: %d\n", epochUnixNano)

	firstOverflow := sort.Search(math.MaxInt64, func(i int) bool {
		//fmt.Printf("trying %d\n", i)
		then := epoch.Add(time.Duration(i) * time.Nanosecond)
		fmt.Printf("trying: %s\n", then.Format(time.RFC3339Nano))
		thenUnixNano := then.UnixNano()
		if thenUnixNano != int64(i) {
			// overlflow
			fmt.Printf("overflowed: %d != %d", thenUnixNano, i)
			return true
		}
		return false
	})
	fmt.Printf("first overflow: %d\n", firstOverflow)

	firstUnderflow := sort.Search(math.MaxInt64, func(i int) bool {
		//fmt.Printf("trying %d\n", -i)
		then := epoch.Add(time.Duration(-i) * time.Nanosecond)
		fmt.Printf("trying: %s\n", then.Format(time.RFC3339Nano))
		thenUnixNano := then.UnixNano()
		if thenUnixNano != int64(-i) {
			// overlflow
			fmt.Printf("underflowed: %d != %d", thenUnixNano, -i)
			return true
		}
		return false
	})
	fmt.Printf("first underflow: %d\n", firstUnderflow)
}
```

But, it actually demonstrates full fidelity of int64 nanoseconds...

Reply:
But it does highlight a real issue, since we use RFC3339 as a our canonical format, we should prefer to use Min/Max values that can be round-tripped safely _in that format_

Reply:
Also, just found these:

https://github.com/blevesearch/bleve/blob/master/document/field_datetime.go#L38-L39

Reply:
Ah great. I could just reuse these settings then:
https://github.com/blevesearch/bleve/blob/d8a53a56c039b7bb29fea965996e6829f4245232/document/field_datetime.go#L38-L39

Let me make sure they work correctly with all the test cases.

Reply:
My concern is that from pure Go this would be fine, but someone mixing Go and JSON might have a problem.  For example, they might pass in MinTimeRepresentable, which is valid, but in a JSON response, we return the RFC3339 represenation of that time, which if later echo'd back in a search request via JSON, it unmarshals to a value that is now out of range (similar to how your example overflowed)...

Reply:
Very valid point, so I might just have to update these entries to the RFC3339 format that I've used in this change.

Reply:
I'm kind of torn on that too.  On the one hand, those existing constants accurately capture the restrictions of our data storage.

Only when querying does this new concern come up, and that happens because we chose RFC3339 as our format.  Another more precise format would not suffer the limitation.

It seems all of these choices might have some backwards compat issues if anyone has data indexed around these edges (unlikely but of course possible)

Reply:
Ugh yea, so I can't reuse those settings for that very reason. You're right - now we're end up turning down some queries that used to go through before, but these queries were quite likely returning some incorrect results due to the overflow.

That said, let me first push up a patch to export these settings and apply them iff QueryDateTimeFormat were RFC3339.

Reply:
On second thought they may not need exporting, as I'll be checking the QueryDateTimeFormat anyway, and this is something the user can choose to set to any other.

Reply:
I want them to be exported, because today users are hard-coding their own incorrect minimums

Reply:
Done.